{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "from collections import defaultdict\n",
    "\n",
    "import aminos"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## setup io"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf = aminos.io.vcf.read_vcf('data/example.vcf.gz')\n",
    "gff = aminos.io.gff.GFF('data/Homo_sapiens.GRCh38.110.chromosome.21.gff3.gz')\n",
    "transcript_reference = aminos.io.fasta.init_transcript_reference('data/reference_sequences.fasta.gz')\n",
    "\n",
    "gff_transcripts = gff.get_unique_transcripts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## iterate through transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/Users/barneyh/brava/aminos/demo.ipynb Cell 5\u001b[0m line \u001b[0;36m1\n\u001b[1;32m    <a href='vscode-notebook-cell:/Users/barneyh/brava/aminos/demo.ipynb#W6sZmlsZQ%3D%3D?line=144'>145</a>\u001b[0m             file\u001b[39m.\u001b[39mwrite(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00mnew_sequence\u001b[39m}\u001b[39;00m\u001b[39m\\n\u001b[39;00m\u001b[39m'\u001b[39m)\n\u001b[1;32m    <a href='vscode-notebook-cell:/Users/barneyh/brava/aminos/demo.ipynb#W6sZmlsZQ%3D%3D?line=146'>147</a>\u001b[0m     file\u001b[39m.\u001b[39mwrite(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m>\u001b[39m\u001b[39m{\u001b[39;00m\u001b[39m\"\u001b[39m\u001b[39m,\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m.\u001b[39mjoin(reference_individuals)\u001b[39m}\u001b[39;00m\u001b[39m\\n\u001b[39;00m\u001b[39m'\u001b[39m)\n\u001b[0;32m--> <a href='vscode-notebook-cell:/Users/barneyh/brava/aminos/demo.ipynb#W6sZmlsZQ%3D%3D?line=147'>148</a>\u001b[0m     file\u001b[39m.\u001b[39mwrite(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00mtranscript_reference\u001b[39m}\u001b[39;00m\u001b[39m\\n\u001b[39;00m\u001b[39m'\u001b[39m)\n\u001b[1;32m    <a href='vscode-notebook-cell:/Users/barneyh/brava/aminos/demo.ipynb#W6sZmlsZQ%3D%3D?line=149'>150</a>\u001b[0m \u001b[39m# write reference:\u001b[39;00m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "mutation_dict = {\n",
    "    \"*frameshift\": \"R\",\n",
    "    \"*frameshift&stop_retained\": \"Q\",\n",
    "    \"*inframe_deletion\": \"C\",\n",
    "    \"*inframe_insertion\": \"J\",\n",
    "    \"*missense\": \"N\",\n",
    "    \"*missense&inframe_altering\": \"K\",\n",
    "    \"*stop_gained\": \"X\",\n",
    "    \"*stop_gained&inframe_altering\": \"A\",\n",
    "    \"frameshift\": \"F\",\n",
    "    \"frameshift&start_lost\": \"V\",\n",
    "    \"frameshift&stop_retained\": \"B\",\n",
    "    \"inframe_deletion\": \"D\",\n",
    "    \"inframe_deletion&stop_retained\": \"P\",\n",
    "    \"inframe_insertion\": \"I\",\n",
    "    \"inframe_insertion&stop_retained\": \"Z\",\n",
    "    \"missense\": \"M\",\n",
    "    \"missense&inframe_altering\": \"Y\",\n",
    "    \"phi\": \"E\",\n",
    "    \"start_lost\": \"0\",\n",
    "    \"start_lost&splice_region\": \"U\",\n",
    "    \"stop_gained\": \"G\",\n",
    "    \"stop_gained&inframe_altering\": \"T\",\n",
    "    \"stop_lost\": \"L\",\n",
    "    \"stop_lost&frameshift\": \"W\"\n",
    "}\n",
    "\n",
    "# take a bcftools csq string and return the transcript and instruction\n",
    "# an instruction is defined as:\n",
    "# (mutation_code, ref_pos, alt_pos, alt_seq, ref_seq_len)\n",
    "\n",
    "def process_csq(csq):\n",
    "    csq = csq.split('|')\n",
    "    \n",
    "    mutation_type = csq[0]\n",
    "    if mutation_type not in mutation_dict:\n",
    "        return None, None\n",
    "\n",
    "    code = mutation_dict[mutation_type]\n",
    "\n",
    "    transcript = csq[2]\n",
    "\n",
    "    match = re.search(r\"(\\d+)([A-Za-z]+)>(\\d+)([A-Za-z]+)\", csq[5])\n",
    "\n",
    "    if match:\n",
    "        ref_pos, ref_seq, alt_pos, alt_seq = match.groups()\n",
    "\n",
    "    else:\n",
    "        return None, None\n",
    "    \n",
    "    return transcript, (code, ref_pos, alt_pos, alt_seq, len(ref_seq))\n",
    "\n",
    "def instructions_to_tasks(instructions_ids, id_to_instruction):\n",
    "    # task is defined as:\n",
    "    # the stream code c, which describes the source of reading data, e.g., reference or\n",
    "    # alteration. Second, the position p, which describes the start of the block in the input stream.\n",
    "    # Third, the length l, which describes the length of the block and finally, the position in the new\n",
    "    # sequence, r\n",
    "\n",
    "    # task (c, p , l, r)\n",
    "    # task (stream_code, position, length, new_position)\n",
    "\n",
    "    tasks = []\n",
    "    current_position = 0  # This represents the current position in the new sequence\n",
    "    \n",
    "    for instruction_id in sorted(instructions_ids):\n",
    "        # Retrieve the instruction from the global dictionary using the ID\n",
    "        instruction = id_to_instruction[instruction_id]\n",
    "        # Parse the instruction string back into a tuple (mutation_code, ref_pos, alt_pos, alt_seq, ref_seq_len)\n",
    "        code, ref_pos, alt_pos, alt_seq, ref_seq_len = eval(instruction)\n",
    "\n",
    "        # Assuming 'R' for reference and 'A' for alteration\n",
    "        # Add a task to copy from the reference up to the current mutation\n",
    "        tasks.append(('R', current_position, int(ref_pos) - current_position, current_position))\n",
    "        # Add a task for the mutation itself\n",
    "        tasks.append(('A', int(alt_pos), len(alt_seq), current_position + (int(ref_pos) - current_position)))\n",
    "        # Update the current position in the new sequence\n",
    "        current_position = int(ref_pos) + len(alt_seq)\n",
    "\n",
    "    tasks.append(('R', current_position, len(transcript_reference), current_position))\n",
    "\n",
    "    return tasks\n",
    "\n",
    "def execute_tasks(reference, tasks):\n",
    "    # execute tasks\n",
    "    # return new sequence\n",
    "    return ...\n",
    "\n",
    "# Initialize a counter for unique instruction IDs\n",
    "instruction_id_counter = 0\n",
    "instruction_to_id = {}\n",
    "\n",
    "def get_instruction_id(instruction):\n",
    "    global instruction_id_counter\n",
    "    # Convert instruction tuple to a string to use as a dictionary key\n",
    "    instruction_key = str(instruction)\n",
    "    if instruction_key not in instruction_to_id:\n",
    "        # Assign a new ID to the instruction if it's not already present\n",
    "        instruction_to_id[instruction_key] = instruction_id_counter\n",
    "        instruction_id_counter += 1\n",
    "    return instruction_to_id[instruction_key]\n",
    "\n",
    "# Process the instructions for each transcript\n",
    "for transcript in gff_transcripts:\n",
    "    instructions = defaultdict(set)\n",
    "\n",
    "    chr, start, end = gff.get_transcript_range(transcript)\n",
    "    for record in vcf(f'{chr}:{start}-{end}'):\n",
    "        bcsq = record.INFO.get('BCSQ')\n",
    "        if bcsq is None:\n",
    "            continue\n",
    "\n",
    "        for csq in bcsq.split(','):\n",
    "            csq_transcript, instruction = process_csq(csq)\n",
    "            if csq_transcript != transcript or not instruction:\n",
    "                continue\n",
    "\n",
    "            instruction_id = get_instruction_id(instruction)\n",
    "            for individual_call, individual in zip(record.genotypes, vcf.samples):\n",
    "                for haplotype in [0, 1]:\n",
    "                    if individual_call[haplotype] == 1:\n",
    "                        instructions[f'{individual}_{haplotype}'].add(instruction_id)\n",
    "\n",
    "    # flip the instruction_to_id dict\n",
    "\n",
    "    id_to_instruction = {id: instruction for instruction, id in instruction_to_id.items()}\n",
    "\n",
    "    with open(f'data/{transcript}.txt', 'w') as file:\n",
    "\n",
    "        reference_individuals = []\n",
    "\n",
    "        for individual in vcf.samples:\n",
    "            for haplotype in [0, 1]:\n",
    "                sample_name = f'{individual}_{haplotype}'\n",
    "\n",
    "                individual_instructions_ids = instructions[sample_name]\n",
    "\n",
    "                if len(individual_instructions_ids) == 0:\n",
    "                    reference_individuals.append(sample_name)\n",
    "            \n",
    "                tasks = instructions_to_tasks(individual_instructions_ids, id_to_instruction)\n",
    "                new_sequence = execute_tasks(transcript_reference, tasks)\n",
    "\n",
    "                file.write(f'>{sample_name}\\n')\n",
    "                file.write(f'{new_sequence}\\n')\n",
    "\n",
    "        file.write(f'>{\",\".join(reference_individuals)}\\n')\n",
    "        file.write(f'{transcript_reference}\\n')\n",
    "                \n",
    "    # write reference:\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "aminos",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
